src/core/lv2/Lv2Proc.cpp

```diff
@@ -0,0 +1,545 @@
+/*
+ * Lv2Proc.cpp - Lv2 processor class
+ *
+ * Copyright (c) 2019-2019 Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2Proc.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include <cmath>
+
+#include "AutomatableModel.h"
+#include "ComboBoxModel.h"
+#include "Engine.h"
+#include "Lv2Manager.h"
+#include "Lv2Ports.h"
+#include "Mixer.h"
+
+
+
+
+Plugin::PluginTypes Lv2Proc::check(const LilvPlugin *plugin,
+	std::vector<PluginIssue>& issues, bool printIssues)
+{
+	unsigned maxPorts = lilv_plugin_get_num_ports(plugin);
+	enum { checkIn, checkOut, checkMax };
+	unsigned audioChannels[checkMax] = { 0, 0 }; // input and output count
+
+	for(unsigned portNum = 0; portNum < maxPorts; ++portNum)
+	{
+		Lv2Ports::Meta meta;
+		// does all port checks:
+		std::vector<PluginIssue> tmp = meta.get(plugin, portNum);
+		std::move(tmp.begin(), tmp.end(), std::back_inserter(issues));
+
+		if(meta.m_type == Lv2Ports::Type::Audio &&
+			!portIsSideChain(plugin,
+							lilv_plugin_get_port_by_index(plugin, portNum)))
+			++audioChannels[meta.m_flow == Lv2Ports::Flow::Output];
+	}
+
+	if(audioChannels[checkIn] > 2)
+		issues.emplace_back(tooManyInputChannels,
+			std::to_string(audioChannels[0]));
+	if(audioChannels[checkOut] == 0)
+		issues.emplace_back(noOutputChannel);
+	else if(audioChannels[checkOut] > 2)
+		issues.emplace_back(tooManyOutputChannels,
+			std::to_string(audioChannels[1]));
+
+	const LilvNodes* reqFeats = lilv_plugin_get_required_features(plugin);
+	LILV_FOREACH(nodes, itr, reqFeats)
```

Comment:
I assume this counts as a for?
```suggestion
	LILV_FOREACH (nodes, itr, reqFeats)
```